src/EventSubscriber/ElasticApmRequestSubscriber.php

```diff
@@ -0,0 +1,238 @@
+<?php
+
+namespace Drupal\elastic_apm\EventSubscriber;
+
+use Drupal\Core\Database\Database;
+use Drupal\Core\Messenger\MessengerInterface;
+use Drupal\Core\Routing\RouteMatchInterface;
+use Drupal\Core\StringTranslation\StringTranslationTrait;
+use Drupal\Core\Utility\Error;
+
+use Drupal\elastic_apm\ElasticApmInterface;
+
+use Exception;
+use Psr\Log\LoggerInterface;
+use Symfony\Component\EventDispatcher\EventSubscriberInterface;
+use Symfony\Component\HttpKernel\Event\FinishRequestEvent;
+use Symfony\Component\HttpKernel\Event\GetResponseEvent;
+use Symfony\Component\HttpKernel\HttpKernelInterface;
+use Symfony\Component\HttpKernel\KernelEvents;
+
+/**
+ * The ElasticApm request event subscriber class.
+ *
+ * @package Drupal\elastic_apm\EventSubscriber
+ */
+class ElasticApmRequestSubscriber implements EventSubscriberInterface {
+
+  use StringTranslationTrait;
+
+  /**
+   * The Elastic APM service object.
+   *
+   * @var \Drupal\elastic_apm\ElasticApmInterface
+   */
+  protected $elasticApm;
+
+  /**
+   * The current path.
+   *
+   * @var \Drupal\Core\Routing\RouteMatchInterface
+   */
+  protected $routeMatch;
+
+  /**
+   * Logger service.
+   *
+   * @var \Psr\Log\LoggerInterface
+   */
+  protected $logger;
+
+  /**
+   * The messenger.
+   *
+   * @var \Drupal\Core\Messenger\MessengerInterface
+   */
+  protected $messenger;
+
+  /**
+   * The actual PHP Agent for the Elastic APM server.
+   *
+   * @var \PhilKra\Agent
+   */
+  protected $phpAgent;
+
+  /**
+   * A flag whether the master request was processed.
+   *
+   * @var bool
+   */
+  protected $processedMasterRequest = FALSE;
+
+  /**
+   * Constructs a ElasticApmRequestSubscriber object.
+   *
+   * @param \Drupal\elastic_apm\ElasticApmInterface $elastic_apm
+   *   The Elastic APM service object.
+   * @param \Drupal\Core\Routing\RouteMatchInterface $route_match
+   *   The current route.
+   * @param \Psr\Log\LoggerInterface $logger
+   *   The logger service.
+   * @param \Drupal\Core\Messenger\MessengerInterface $messenger
+   *   The messenger.
+   */
+  public function __construct(
+    ElasticApmInterface $elastic_apm,
+    RouteMatchInterface $route_match,
+    LoggerInterface $logger,
+    MessengerInterface $messenger
+  ) {
+    $this->elasticApm = $elastic_apm;
+    $this->routeMatch = $route_match;
+    $this->logger = $logger;
+    $this->messenger = $messenger;
+
+    // Initialize the PHP agent if the Elastic APM config is configured.
+    if ($this->elasticApm->isConfigured()) {
+      $this->phpAgent = $this->elasticApm->getAgent();
+    }
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public static function getSubscribedEvents() {
```

Comment:
**CHANGE REQUEST**

*Performance*

Can you try the following and see if it works.
Check if Elastic APM is enabled AND configured, and if not return without registering the events. That will prevent the subscriber from listening in the first place instead of listening but doing nothing.

*Error logging*

Log an error if it is enabled but not configured.